crates/game/src/world/interactions/instruments/compose.rs

```diff
@@ -0,0 +1,147 @@
+use std::collections::HashSet;
+use strum::IntoEnumIterator;
+
+use common::entities::{Identity, Instrument, Song, VocalStyle};
+use simulation::queries::character::SkillQueries;
+
+use crate::constants;
+use crate::context::Context;
+use crate::world::interactions::*;
+
+#[derive(Clone, Default)]
+pub struct ComposeInteraction {
+    pub instrument: Instrument,
+}
+
+impl Identity for ComposeInteraction {
+    fn id(&self) -> String {
+        "compose".into()
+    }
+}
+
+impl Interaction for ComposeInteraction {
+    fn name(&self) -> String {
+        "Compose".into()
+    }
+
+    fn description(&self) -> String {
+        "Composing will allow you to create song ideas or improved previous ideas that you can later record".into()
+    }
+
+    fn requirements(&self) -> Vec<Requirement> {
+        vec![Requirement::HealthAbove(20), Requirement::MoodAbove(20)]
+    }
+
+    fn track_action(&self) -> bool {
+        true
+    }
+
+    fn limit_daily_interactions(&self) -> InteractionTimes {
+        InteractionTimes::Multiple(2)
+    }
+
+    fn effects(&self, context: &Context) -> InteractionEffects {
+        let instrument_skill = context
+            .game_state
+            .character
+            .get_skill_with_level(&self.instrument.associated_skill);
+
+        // Round up to 1 if it's zero, otherwise the song won't improve.
+        let instrument_skill_level = if instrument_skill.level == 0 {
+            1
+        } else {
+            instrument_skill.level
+        };
+
+        InteractionEffects {
+            always_applied: vec![
+                InteractionEffect::Energy(EffectType::Negative(
+                    constants::effects::negative::HEALTH_NORMAL_INTERACTION,
+                )),
+                InteractionEffect::Time(TimeConsumption::TimeUnit(2)),
+            ],
+            applied_after_interaction: vec![InteractionEffect::Song(EffectType::Positive(
+                constants::effects::positive::SONG_COMPOSE_INTERACTION * instrument_skill_level,
+            ))],
+        }
+    }
+
+    fn sequence(&self, context: &Context) -> InteractSequence {
+        let songs_in_progress = context.clone().game_state.character.songs_in_progress;
+        if !songs_in_progress.is_empty() {
+            Ok(build_existing_song_sequence(songs_in_progress, context))
+        } else {
+            Ok(build_new_song_sequence(context))
+        }
+    }
+
+    fn messages(&self, _context: &Context) -> (String, String) {
```

Comment:
This way of passing message does not seem correct. In here we should be able to show the name of the song that was improved and problably do some pretty printing on the CLI side. Maybe storing messages here is not a good idea.

Reply:
Moving this into an enum, something like a `InteractionOutcome` and consumed by the CLI would be cool.